### PR TITLE
Create some unit tests for Doc syntax

### PR DESCRIPTION
**Pull request by sellout**
## Overview

The goal of this is faster testing cycles and more precise test cases than using transcripts. However, there is a bunch of background work to make it work. Most notably it

- creates a `unison-util-recursion` package to unify the recursion-schemey code around the repo and
- restructures the Doc data types to allow for more consistent annotation (using Cofree), so that it can be filtered out for the tests (like how `Token` is removed from the lexer tests).

## Implementation notes

This takes more advantage of fixed-point operators to allow changing the structure more easily (dropping annotations for test cases).

The Doc restructuring also encodes more of the restrictions in types (which had previously been in comments). We don’t want to keep those restrictions, but they reflect what the parser is currently able to parse.

This PR also changes the lexer unit tests a bit to catch some failures that could previously be missed. E.g., since it dropped the first and last token of the actual result (to match the provided expected result), an `Err` lexeme could be dropped, making a failure look like a success when compared against an empty expected result, or at least obscuring the reason for a failure.

## Interesting/controversial decisions

I don’t know … recursion schemes? Although they’re already used in a few places, like the ABTs.

## Test coverage

This adds a bunch of new unit tests, but none of them are particularly interesting, just a few to justify the existence of the new test suite.

## Loose ends

Moar tests in the suite, that focus on edge cases. But with upcoming changes like #5255, it seems premature to try to explore the boundaries of the current syntax.

### CODE REVIEW

**Review by aryairani — APPROVED** [review]

File: lib/unison-util-recursion/src/Unison/Util/Recursion.hs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+deriving instance (forall a. (Eq a) => Eq (f a)) => Eq (Fix f)
+
+deriving instance (Eq (Fix f), forall a. (Ord a) => Ord (f a)) => Ord (Fix f)
+
+instance (Functor f) => Recursive (Fix f) f where
+  embed = Fix
+  project (Fix f) = f
+
+data Cofree' f a x = a :<< f x
+  deriving (Foldable, Functor, Traversable)
```
Comment (ChrisPenner) [inline lib/unison-util-recursion/src/Unison/Util/Recursion.hs:50]: Just curious, any reason why not use `CofreeF` from `free`?

File: lib/unison-util-recursion/src/Unison/Util/Recursion.hs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  )
+where
+
+import Control.Arrow ((&&&))
+import Control.Comonad.Cofree (Cofree ((:<)))
+import Control.Monad ((<=<))
+
+type Algebra f a = f a -> a
+
+class Recursive t f | t -> f where
```
Comment (ChrisPenner) [inline lib/unison-util-recursion/src/Unison/Util/Recursion.hs:21]: Just curious why not use `recursion-schemes` directly?

I personally find your encoding here less confusing than all the `Base` type-family nonsense, but it's also nice to maintain compatibility with the ecosystem where possible 😄

File: lib/unison-util-recursion/src/Unison/Util/Recursion.hs (hunk 1 above)
Comment (sellout) [reply lib/unison-util-recursion/src/Unison/Util/Recursion.hs:50]: I should have. I think I got confused and tried to use the isomorphic `EnvT` … but that’s from transformers, and I didn’t find it in free (of course), so I was like “oh, maybe the pattern functor isn’t here.”

In my own code, I don’t use those because they’re not strict in the functor parameter, so they don’t result in lawful fixed points. But my `Cofree'` here isn’t strict either, so …

File: lib/unison-util-recursion/src/Unison/Util/Recursion.hs (hunk 2 above)
Comment (sellout) [reply lib/unison-util-recursion/src/Unison/Util/Recursion.hs:21]: Yeah, my plan was to replace this with a recursion scheme library eventually. At this point, I was just trying to centralize the (mostly) existing copies of this pattern.

I’m happy to put in recursion-schemes, but I also have my own recursion scheme library, [Yaya](https://hackage.haskell.org/package/yaya), which makes some different decisions than recursion-schemes[^1]. One of them is the use of `FunctionalDependencies`, as you pointed out. Another is that Yaya tries to provide totality (especially when paired with [`NoRecursion`](https://hackage.haskell.org/package/no-recursion)).

So I just punted on that discussion for now.

[^1]: There’s a more in-depth comparison with recursion-schemes in [Yaya’s README](https://github.com/sellout/yaya/blob/main/core/README.md#recursion-schemes).

File: lib/unison-util-recursion/src/Unison/Util/Recursion.hs (hunk 1 above)
Comment (sellout) [reply lib/unison-util-recursion/src/Unison/Util/Recursion.hs:50]: This is addressed by #5500.
